aip/0160.md

```diff
@@ -0,0 +1,166 @@
+---
+aip:
+  id: 160
+  state: reviewing
+  created: 2020-02-24
+permalink: /160
+redirect_from:
+  - /0160
+---
+
+# Filtering
+
+Often, when listing resources (using a list method as defined in [AIP-132][] or
+something reasonably similar), it is desirable to filter over the collection
+and only return results that the user is interested in.
+
+It is tempting to define a structure to handle the precise filtering needs for
+each API. However, filtering requirements evolve frequently, and therefore it
+is prudent to use a string field with a structured syntax accessible to a
+non-technical audience. This allows updates to be able to be made
+transparently, without waiting for UI or client updates.
+
+**Note:** Because list filters are intended for a potentially non-technical
+audience, they sometimes borrow from patterns of colloquial speech rather than
+common patterns found in code.
+
+## Guidance
+
+APIs **may** provide filtering to users on `List` methods. If they choose to do
+so, they **must** follow the common specification for filters discussed here.
+The syntax is formally defined in the [EBNF grammar][].
+
+**Note:** Google provides implementations for the filtering language in both
+[C++][cel-cpp] and [Go][cel-go] as a discrete component of [CEL][]. The use of
+these implementations is encouraged.
+
+### Expressions
+
+Expressions are literal values in the filtering syntax ("42", "Victor Hugo").
+Expressions can be pieced together with operators to make other expressions.
+
+### Logical Operators
+
+Filtering implementations **must** provide the binary operators `AND` and `OR`,
+as well as a whitespace operator:
+
+| Operator | Example       | Meaning                                |
+| -------- | ------------- | -------------------------------------- |
+| `AND`    | `a AND b`     | True if `a` and `b` are true.          |
+| `OR`     | `a OR b OR c` | True if any of `a`, `b`, `c` are true. |
+
+The whitespace operator (example: `a b c`) represents a fuzzy variant of `AND`,
+which is suitable for result ranking and term rewrite.
+
+**Note:** To match common patterns of speech, the `OR` operator has higher
+precedence than `AND`, unlike what is found in most programming languages. The
+expression `a AND b OR c` evaluates: `a AND (b OR c)`. API documentation and
+examples **should** encourage the use of explicit parentheses to avoid
+confusion.
+
+### Negation Operators
+
+Filtering implementations **must** provide the unary operators `NOT` and `-`.
+These are used interchangeably.
+
+| Operator | Example | Meaning                  |
+| -------- | ------- | ------------------------ |
+| `NOT`    | `NOT a` | True if `a` is not true. |
+| `-`      | `-a`    | True if `a` is not true. |
+
+### Comparison Operators
+
+Filtering implementations **must** provide the binary comparison operators `=`,
+`!=`, `<`, `>`, `<=`, and `>=`.
+
+| Operator | Example      | Meaning                                         |
```

Comment:
I originally wrote a separate table for bool, number, and string, but I collapsed it because it was getting quite verbose.

We _could_ bring over the JavaScript for tabbed sections (already done in #388) though, which might work well if this is deemed to be too terse. (In particular, it might help call out things like the `*` syntax for strings.)

Reply:
I think this is okay as-is. The wildcarding before and after for strings is pretty special case.